src/analyses/loopTermination.ml

```diff
@@ -0,0 +1,109 @@
+(** Termination analysis for loops and [goto] statements ([termination]). *)
+
+open Analyses
+open GoblintCil
+open TerminationPreprocessing
+
+(** Stores the result of the query whether the program is single threaded or not
+    since finalize does not have access to ctx. *)
+let single_thread : bool ref = ref false
+
+(** Contains all loop counter variables (varinfo) and maps them to their corresponding loop statement. *)
+let loop_counters : stmt VarToStmt.t ref = ref VarToStmt.empty
+
+(** Checks whether a variable can be bounded. *)
+let check_bounded ctx varinfo =
+  let open IntDomain.IntDomTuple in
+  let exp = Lval (Var varinfo, NoOffset) in
+  match ctx.ask (EvalInt exp) with
+  | `Top -> false
+  | `Lifted v -> not (is_top_of (ikind v) v)
+  | `Bot -> failwith "Loop counter variable is Bot."
+
+(** We want to record termination information of loops and use the loop
+ * statements for that. We use this lifting because we need to have a
+ * lattice. *)
+module Statements = Lattice.Flat (CilType.Stmt) (Printable.DefaultNames)
+
+(** The termination analysis considering loops and gotos *)
+module Spec : Analyses.MCPSpec =
+struct
+
+  include Analyses.IdentitySpec
+
+  let name () = "termination"
+
+  module D = Lattice.Unit
+  module C = D
+  module V =
+  struct
+    include Printable.Unit
+    let is_write_only _ = true
+  end
+  module G = MapDomain.MapBot (Statements) (BoolDomain.MustBool)
+
+  let startstate _ = ()
+  let exitstate = startstate
+
+  (** Warnings for detected possible non-termination *)
+  let finalize () =
+    (* Multithreaded *)
+    if not (!single_thread) then (
+      M.warn ~category:NonTerminating "The program might not terminate! (Multithreaded)\n"
+    )
+
+  (** Recognizes a call of [__goblint_bounded] to check the EvalInt of the
+   * respective loop counter variable at that position. *)
+  let special ctx (lval : lval option) (f : varinfo) (arglist : exp list) =
+    if !AnalysisState.postsolving then
+      match f.vname, arglist with
+        "__goblint_bounded", [Lval (Var x, NoOffset)] ->
+        (try
+           let loop_statement = VarToStmt.find x !loop_counters in
+           let is_bounded = check_bounded ctx x in
+           ctx.sideg () (G.add (`Lifted loop_statement) is_bounded (ctx.global ()));
+           (* In case the loop is not bounded, a warning is created. *)
+           if not (is_bounded) then (
+             let msgs =
+               [(Pretty.dprintf
+                   "The program might not terminate! (Loop analysis)",
+                 Some (M.Location.CilLocation (Cilfacade.get_stmtLoc loop_statement))
+                );] in
+             M.msg_group Warning ~category:NonTerminating "Possibly non terminating loops" msgs);
+           ()
+         with Not_found ->
+           (* This should not happen as long as __goblint_bounded is only used
+            * for this analysis. *)
```

Comment:
The `Not_found` can be raised by `VarToStmt.find x !loop_counters` in line 62 in case we encounter a call of `__goblint_bounded` where the `x` in the argument `[Lval (Var x, NoOffset)]` is not registered as a loop counter variable.

Reply:
Yes, but the comment says it should never happen in the first place, unless there's some bug. Currently this would hide the bug, but it would be better to just not catch it at all. If there's ever an issue, Goblint will crash and we'll know that something with `loop_counters` is buggy.